### PR TITLE
Improve performance of LitPop

### PR DESCRIPTION
**Pull request by peanutfun**
Changes proposed in this PR:
- Speed up country-code detection in `LitPop` by assuming the data is gridded (it is)
- Streamline `gridpoints_core_calc` to use numpy array tools and broadcasting, and avoid copying arrays
- Add option to pass region ID to `LitPop.from_shape`. This avoids country-code detection.

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [x] Source branch up-to-date with target branch
- [ ] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [x] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

### CODE REVIEW

File: climada/entity/exposures/litpop/litpop.py [hunk 1]
```diff
@@ -755,18 +774,19 @@ def _get_litpop_single_polygon(polygon, reference_year, res_arcsec, data_dir,
...
-                                     np.round_(lon.flatten(), decimals = 8, out = None),
-                                     np.round_(lat.flatten(), decimals = 8, out = None)
-                                     )
-                                 )
-    gdf['latitude'] = np.round_(lat.flatten(), decimals = 8, out = None)
-    gdf['longitude'] = np.round_(lon.flatten(), decimals = 8, out = None)
+    latitude = np.round(lat.flatten(), decimals=8)
+    longitude = np.round(lon.flatten(), decimals=8)
+    gdf = geopandas.GeoDataFrame(
+        {'value': litpop_array.flatten(), "latitude": latitude, "longitude": longitude},
```
Comment (peanutfun) [inline climada/entity/exposures/litpop/litpop.py:780]: Normalize string delimiters
```suggestion
        {"value": litpop_array.flatten(), "latitude": latitude, "longitude": longitude},
```

File: climada/entity/exposures/litpop/litpop.py [hunk 2]
```diff
@@ -992,47 +1014,49 @@ def gridpoints_core_calc(data_arrays, offsets=None, exponents=None,
 
     Returns
     -------
-    result_array : np.array of same shape as arrays in data_arrays
+    result : np.asarray of same shape as arrays in data_arrays
```
Comment (peanutfun) [inline climada/entity/exposures/litpop/litpop.py:1017]:
```suggestion
    result : np.array of same shape as arrays in data_arrays
```

File: climada/entity/exposures/litpop/litpop.py [hunk 3]
```diff
@@ -488,6 +498,9 @@ def from_shape(cls, shape, total_value, res_arcsec=30, exponents=(1,1), value_un
             Defining power with which lit (nightlights) and pop (gpw) go into LitPop.
         value_unit : str
             Unit of exposure values. The default is USD.
+        region_id : int, optional
+            The numeric ISO 3166 region associated with the shape. If ``None``, it will
+            be determined automatically (at a slight computational cost).
```
Comment (chahank) [inline climada/entity/exposures/litpop/litpop.py:503]: I think this wording is slightly confusing. I thought `region_id` would be an argument to select the shape automatically. But, as I understand from this docstring, it is just the `region_id` that will be assigned to all grid points within the shape? How do you handle shapes that go over multiple regions?

File: climada/entity/exposures/litpop/litpop.py [hunk 4]
```diff
@@ -992,47 +1014,49 @@ def gridpoints_core_calc(data_arrays, offsets=None, exponents=None,
...
+        """Prepare the offset and exposure inputs
+
+        If they are ``None``, use the default values. If not, cast them to an array and
+        ensure that the array has only one dimension. Finally, check if the values are
+        positive.
+        """
+        if input_data is None:
+            return default_value
+
+        ret = np.asarray(input_data, dtype=float)
```
Comment (chahank) [inline climada/entity/exposures/litpop/litpop.py:1034]: Curiosity: what is the variable name `ret` supposed to mean?

**Comment by chahank** [comment]
Looks all very reasonable! Any estimates of the gained computation time? Also, changelog needs to be updated. Maybe also the documentation to point people towards how to use custom shapes correctly.

File: climada/entity/exposures/litpop/litpop.py (hunk 3 above)
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py:503]: Well, I can only tell you what is implemented at the moment. You can input a single shape into `from_shape`. This will call `_get_litpop_single_polygon` once, and in this function, the `region_id` is either set to the input value or determined automatically. See here: https://github.com/CLIMADA-project/climada_python/blob/509326234c49e7534e7a9a3d46e35425d70499f0/climada/entity/exposures/litpop/litpop.py#L784

I just added the option to pipe the `region_id` through `from_shape`.

> But, as I understand from this docstring, it is just the region_id that will be assigned to all grid points within the shape?

This is what `_get_litpop_single_polygon` has been doing and still does, if you pass `region_id` to it.

> How do you handle shapes that go over multiple regions?

You either let the `region_id` be set automatically, or you choose a single ID for the entire shape (which might actually be something you want?)

File: climada/entity/exposures/litpop/litpop.py [hunk 5]
```diff
@@ -512,10 +525,15 @@ def from_shape(cls, shape, total_value, res_arcsec=30, exponents=(1,1), value_un
...
-                                                          gpw_version, exponents,
-                                                          )
+        litpop_gdf, _ = _get_litpop_single_polygon(
+            shape,
+            reference_year,
+            res_arcsec,
+            data_dir,
+            gpw_version,
+            exponents,
+            region_id=region_id,
```
[the remark this answers is not in the document]
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py:535]: Nope, I just wasn't sure about the parameter ordering. But indeed, using the kwarg can be removed.
```suggestion
            region_id,
```

File: climada/entity/exposures/litpop/litpop.py [hunk 6]
```diff
@@ -673,8 +691,9 @@ def _get_litpop_single_polygon(polygon, reference_year, res_arcsec, data_dir,
         Defining power with which lit (nightlights) and pop (gpw) go into LitPop. To get
         nightlights^3 without population count: (3, 0). To use population count alone: (0, 1).
     region_id : int, optional
-        if provided, region_id of gdf is set to value.
-        The default is None, this implies that region_id is not set.
+        if provided, region_id of gdf is set to value. The default is ``None``, in which
+        case the region is determined automatically for every location of the resulting
+        GeoDataFrame.
```
[the remark this answers is not in the document]
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py:696]: But it is exactlty what the function does 😅

File: climada/entity/exposures/litpop/litpop.py (hunk 4 above)
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py:1034]: It's a pattern I personally stick to. I like to call the returned variable `ret`, for `return_value` or something. It's short and makes sense to me when reading `return ret`. You could also call it `result`, `res`, whatever.

**Comment by peanutfun** [comment]
> Any estimates of the gained computation time?

I was afraid you would ask that 😬  I can come up with some benchmarks tomorrow. The most important improvement is from using `get_country_code(gridded=True)`. The actual LitPop computations are very fast, and avoiding copies does not actually help much with that. In my use case, LitPop was occupied for some 50s, 90% of which were spent with `get_country_code`. The other large bottleneck I found was `geopandas.points_from_xy`, but that resolved itself after upgrading from shapely v1.8 to v2.0.

So, tldr: `from_shape` can be by a factor 10 faster with this PR. But I'll run a benchmark just to be sure.

**Comment by peanutfun** [comment]
> Maybe also the documentation to point people towards how to use custom shapes correctly.

Not sure what you mean by that. How are people using it "incorrectly" in your perspective?

File: climada/entity/exposures/litpop/litpop.py (hunk 3 above)
Comment (chahank) [reply climada/entity/exposures/litpop/litpop.py:503]: I see. Then we just let it as is. I maybe would just modify a bit the docstring wording and make it a bit more verbose (at least for the exposed methods).

File: climada/entity/exposures/litpop/litpop.py (hunk 4 above)
Comment (chahank) [reply climada/entity/exposures/litpop/litpop.py:1034]: I think `ret` (as is `result` or `res`) is then a sub-optimal naming. It is important to understand what you are returning, not that you are returning something imho. While the structure of the code is easy to understand with `ret`, its purpose is obscured.

**Comment by chahank** [comment]
> Not sure what you mean by that. How are people using it "incorrectly" in your perspective?

Sorry for being unclear. I was referring to the original issue which was about how to use arbitrary shapes with LitPop. This functionality was clearly not obvious, so a documentation entry could help.

**Comment by peanutfun** [comment]
Ah, sorry, you were referring to #715. I'll see what I can do.

File: climada/entity/exposures/litpop/litpop.py (hunk 4 above)
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py:1034]: Point taken, will think of something to improve this here

**Comment by peanutfun** [comment]
So here are some benchmark results.

### `LitPop.from_countries`
(basically no change)

on develop
Screenshot 2023-05-16 at 12 30 12

on this branch
Screenshot 2023-05-16 at 12 27 38


### `LitPop.from_shape`
(time spent in `get_country_code` dropped from 48% to 13 %)

on develop
Screenshot 2023-05-16 at 12 29 52


on this branch

Screenshot 2023-05-16 at 12 27 07

File: climada/entity/exposures/litpop/litpop.py (hunk 3 above)
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py]: Done, see https://github.com/CLIMADA-project/climada_python/pull/720/commits/bf5d8c9ab621fe1445bac9a5998cfaea05657655

File: climada/entity/exposures/litpop/litpop.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py]: Done, see https://github.com/CLIMADA-project/climada_python/pull/720/commits/bf5d8c9ab621fe1445bac9a5998cfaea05657655

File: climada/entity/exposures/litpop/litpop.py [hunk 7]
```diff
@@ -992,47 +1017,49 @@ def gridpoints_core_calc(data_arrays, offsets=None, exponents=None,
 
     Returns
     -------
-    result_array : np.array of same shape as arrays in data_arrays
+    result : np.array of same shape as arrays in data_arrays
```
Comment (chahank) [inline climada/entity/exposures/litpop/litpop.py:1020]: This is really nitpicking, but `result` is a terrible variable name ^^.

File: climada/entity/exposures/litpop/litpop.py [hunk 8]
```diff
@@ -992,47 +1017,49 @@ def gridpoints_core_calc(data_arrays, offsets=None, exponents=None,
...
+
+    def prepare_input(input_data, default_value, name):
+        """Prepare the offset and exposure inputs
+
+        If they are ``None``, use the default values. If not, cast them to an array and
+        ensure that the array has only one dimension. Finally, check if the values are
+        positive.
+        """
+        if input_data is None:
+            return default_value
```
Comment (chahank) [inline climada/entity/exposures/litpop/litpop.py:1035]: Is it intended that the function returns different types depending on the inputs? (np.array or whatever type is `default_value`)

**Comment by peanutfun** [comment]
> Small improvements, but certainly useful!

Yes, I think I mixed something up. The major improvement was upgrading to shapely v2.0 on my side.

File: climada/entity/exposures/litpop/litpop.py (hunk 8 above)
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py:1035]: Yes. The returned value is used in numpy array calculations, where the usual broadcasting behavior applies. So we are fine with returning scalars or arrays.

File: climada/entity/exposures/litpop/litpop.py (hunk 7 above)
Comment (chahank) [reply climada/entity/exposures/litpop/litpop.py:1020]: Also fine. Although `data` is not a better name than `result`.

File: climada/entity/exposures/litpop/litpop.py (hunk 8 above)
Comment (chahank) [reply climada/entity/exposures/litpop/litpop.py:1035]: I personally find it confusing, but it has a very limited scope so we can let it be.

File: climada/entity/exposures/litpop/litpop.py (hunk 7 above)
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py:1020]: Then I'll do that. My goal is not to refactor this function 😅

File: climada/entity/exposures/litpop/litpop.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py:699]: Would you consider this resolved with the update of the docstring to `from_shape`?

File: climada/entity/exposures/litpop/litpop.py (hunk 7 above)
Comment (peanutfun) [reply climada/entity/exposures/litpop/litpop.py]: Done in 620d3460f9fc4e159fd2ced3aecda705e5bd2fb6

**Comment by peanutfun** [comment]
@chahank Ready to merge from my side!
